### DISCUSSION TITLE
CANNOT RENDER HTML IN FASTAPI

### DISCUSSION THREAD

**Discussion by siamhassan66** [fastapi/fastapi#7225] [Questions]
I am a beginner in python and I am trying to build an app in fastapi.

My task is to get text from html,
render html in fastpi
run python code on the text
and return the output to html page

This is my code

`from typing import Optional

from fastapi import FastAPI, Request, Response
from fastapi.staticfiles import StaticFiles
from fastapi.responses import HTMLResponse
from fastapi.templating import Jinja2Templates,

app = FastAPI()

app.mount("/static", StaticFiles(directory="static"), name="static")

templates = Jinja2Templates(directory="templates")

@app.get("/", response_class=HTMLResponse)
async def get():
    return templates.TemplateResponse("layout.html")

@app.post("/")
async def post(request: Request):
    text = request.get("sent")
`

**Comment by Kludex** [comment]
WE'D LOVE TO HELP YOU! :peace_symbol: 

But we need your help first... :runner:

Can you:
- Format your code snippet:
   - Image:
![image](https://user-images.githubusercontent.com/7353520/89934213-88be9400-dc10-11ea-9a59-b4fbb4640ed7.png)
   - Generated code:
```python
from fastapi import FastAPI

app = FastAPI()
```
- Be specific on how we can help you.
   - What is the problem?
   - What is failing?
   - What do you want to achieve?
- Last but the most important: please, follow our [question issue template](https://github.com/tiangolo/fastapi/blob/master/.github/ISSUE_TEMPLATE/question.md). It makes us know where you have searched possible solutions, understand your issue better and several benefits for those who have the same problem as you and for those who are trying to help as well. :sunglasses:

**Comment by ycd** [comment]
You have to pass the request as part of the key-value pairs in the context for Jinja2.  


The fix is quite easy you just need to add Request in query parameters, 


```
from fastapi import Request
...
@app.get("/", response_class=HTMLResponse)
async def get(request: Request):
    return templates.TemplateResponse("layout.html", {"request": request})
```
Then pass all the data inside the dictionary that you want to render.

**Comment by siamhassan66** [comment]
@Kludex  @ycd 


Code:

```
import json

from fastapi import FastAPI, Request
from fastapi.staticfiles import StaticFiles
from fastapi.responses import HTMLResponse
from fastapi.templating import Jinja2Templates
from urduhack.models.ner.predict import predict_ner


app = FastAPI()

app.mount("/static", StaticFiles(directory="static"), name="static")

templates = Jinja2Templates(directory="templates")


@app.get("/", response_class=HTMLResponse)
async def get(request: Request):
    return templates.TemplateResponse("layout.html", {"request": request})


@app.post("/", response_model=HTMLResponse)
async def post(request: Request):
    text = request.get("sentence")
    entities = predict_ner(text)
    if entities:
        return json.dumps(dict(entities))
```


This is what I can put forward here,

Like in Flask there is a function, render_template , it renders whole html page and we can select the things we want.


I want to do the same thing, get method renders html page but I can't use it in the post method, it returns null. Can you guide me on that, as no documentation is available from fastapi.

**Comment by siamhassan66** [comment]
@Kludex if you still don't understand, below is the link to kind of thing I want to build using fastapi

[anago-tagger](https://github.com/vunb/anago-tagger/tree/master/anago/server)

**Comment by siamhassan66** [comment]
> You have to pass the request as part of the key-value pairs in the context for Jinja2.
> 
> The fix is quite easy you just need to add Request in query parameters,
> 
> ```
> from fastapi import Request
> ...
> @app.get("/", response_class=HTMLResponse)
> async def get(request: Request):
>     return templates.TemplateResponse("layout.html", {"request": request})
> ```
> 
> Then pass all the data inside the dictionary that you want to render.

I think you understood my point, 
Can you please elaborate a bit on that point, 
Let's say I want to select a div text whose id is "sentence",

```
def post():
how can I get the text here, run a machine learning model and then return it
```

**Comment by ycd** [comment]
You should be good to go with the same way i mentioned above


`return templates.TemplateResponse("layout.html", {"request": request, "data": data})`


Now you successfully rendered, you just need to use that form front-end

```
<body>
    <div class="list-group">
          <div class="d-flex w-100 justify-content-between">
             {% for value in data %}
            <h5 class="mb-1">{{ value.title }}</h5>
             {% endfor %}
          </div>
     </div>
</body>
```


> Let's say I want to select a div text whose id is "sentence"

I didn't understand that so you want to scrape a page and render it ?

**Comment by siamhassan66** [comment]
> You should be good to go with the same way i mentioned above
> 
> `return templates.TemplateResponse("layout.html", {"request": request, "data": data})`
> 
> Now you successfully rendered, you just need to use that form front-end
> 
> ```
> <body>
>     <div class="list-group">
>           <div class="d-flex w-100 justify-content-between">
>              {% for value in data %}
>             <h5 class="mb-1">{{ value.title }}</h5>
>              {% endfor %}
>           </div>
>      </div>
> </body>
> ```
> 
> > Let's say I want to select a div text whose id is "sentence"
> 
> I didn't understand that so you want to scrape a page and render it ?

@ycd thanks for your response,

here is y layout.html page,

```
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>AnaGo-demo</title>
    <meta http-equiv="x-ua-compatible" content="ie=edge">
    <meta name="MobileOptimized" content="320">
    <meta name="description" content="">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/bulma/0.4.2/css/bulma.min.css">
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.css">
    <script src="https://code.jquery.com/jquery-3.1.1.min.js" integrity="sha256-hVVnYaiADRTO2PzUGmuLJr8BLUSjGIZsDYGmIJLv2b8=" crossorigin="anonymous"></script>
    <script src="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.js"></script><link rel="stylesheet" type="text/css" href="mystyle.css">
    <link href="{{ url_for('static', path='/style.css') }}" rel="stylesheet">
</head>
<body>
<div id="app" class="container">
    <div class="content-area">
        <div class="columns">
            <div class="column">
                <div class="field">
                    <label class="label">Sentence</label>
                    <div class="control">
                        <form action="/" method="post">
                            <textarea id="sentence" name="sentence" class="textarea" placeholder="Enter English Text Here"></textarea><br>
                            <textarea id="ip" name="ip" class="textarea"></textarea>
                        </form>
                    </div>
            </div>
                <label class="label">Named Entities</label>
                <div class="example">
                    <div class="content">
                        <p>
                            <small class="axlabel organization">Organization</small>
                            <small class="axlabel location">Location</small>
                            <small class="axlabel person">Person</small>
                            <small class="axlabel misc">Misc</small>
                        </p>
                    </div>
                </div>
        </div>
    </div>
</div>
    {% block content %}
    {% endblock %}

    </div>
    <script>
       ws= $('#sentence').on('input', function(e){
        var d= e.target.value;
        $('#ip').val(d);
        })
    </script>

</body>

</html>
```

and I want to select text from

```
<textarea id="sentence" name="sentence" class="textarea" placeholder="Enter English Text Here"></textarea>
```

How can I do that with Fastapi. Can you please guide me on that ?

**Comment by siamhassan66** [comment]
> You should be good to go with the same way i mentioned above
> 
> `return templates.TemplateResponse("layout.html", {"request": request, "data": data})`
> 
> Now you successfully rendered, you just need to use that form front-end
> 
> ```
> <body>
>     <div class="list-group">
>           <div class="d-flex w-100 justify-content-between">
>              {% for value in data %}
>             <h5 class="mb-1">{{ value.title }}</h5>
>              {% endfor %}
>           </div>
>      </div>
> </body>
> ```
> 
> > Let's say I want to select a div text whose id is "sentence"
> 
> I didn't understand that so you want to scrape a page and render it ?

@ycd I am not familiar with difference in scraping and rendering in terms of web

**Comment by ycd** [comment]
I'm not a frontend guy but In this case you need to send a request from Front-end to Back-end, with Axios or Ajax etc. I only know vue and axios so i created a example for you with axios and vue.


**home.html**

```javascript
<html>
<title></title>
<script src="https://cdn.jsdelivr.net/npm/vue/dist/vue.js"></script>
<script src="https://unpkg.com/axios/dist/axios.min.js"></script>

<body>
    <div id="app">
        <textarea name="" id="content" cols="30" rows="10" v-model="content"></textarea>
        <button @click="addText" id="add-textarea">click me</button>
    </div>

    <script>
        new Vue({
            el: "#app",
            data: {
                title: '',
                content: ''
            },
            methods: {
                addText() {
                    return axios.post("/add", {
                        content: this.content
                    }, {
                        headers: {
                            'Content-type': 'application/json',
                        }
                    }).then((response) => {
                        console.log("content: " + this.content);
                    });
                }
            }
        });
    </script>
</body>

</html>
```

**main.py**

```python
from fastapi import FastAPI, Request
from fastapi.templating import Jinja2Templates
from pydantic import BaseModel

templates = Jinja2Templates(directory="templates")

app = FastAPI()


class TextArea(BaseModel):
    content: str


@app.post("/add")
async def post_textarea(data: TextArea):
    print(data.dict())
    return {**data.dict()}


@app.get("/")
async def serve_home(request: Request):
    return templates.TemplateResponse("home.html", {"request": request})
```

When i go to my root path:

![image](https://user-images.githubusercontent.com/62724709/90125078-0f3ab900-dd6a-11ea-8979-33c017305a18.png)

I'm seeing this textarea and click button, now let's send a request.


```
 {'content': 'Hello textarea!'}
INFO:     127.0.0.1:51682 - "POST /add HTTP/1.1" 200 OK
```

**Comment by ycd** [comment]
Is the problem resolved @siamhassan66 ?

**Comment by siamhassan66** [comment]
not yet

**Comment by eliteuser26** [comment]
I went through the same problem with the post request. I tend to agree that the Fastapi documentation doesn't explain how to link the post request with an html web page. It shows what you need in Fastapi only without the other part. I figured out how to do it by myself. I asked that question too but received no answer for it. There is a need of an example to do this in the documentation. I can contribute on this aspect. I generated web pages with fastapi and it works similar to Flask. Had to relearn to program it in Fastapi code.

**Comment by ycd** [comment]
Is the problem resolved @siamhassan66 ? If it is, do you mind closing the issue?

**Comment by Barnez299** [comment]
I noticed this issue still open. I logged a similar one [https://github.com/tiangolo/fastapi/issues/2525](url)
@eliteuser26 please review and let me know. SO link here--[https://stackoverflow.com/questions/65335763/fastapi-render-template-in-index-html-not-working](url)

**Comment by CakJuice** [comment]
For fastapi & jinja2 templates, you should use absolute path,
for example:

main.py

```
from pathfile import Path

# some of codes here....

BASE_PATH = Path(__file__).parent.resolve()
# lets say your template directory is under the root directory
templates = Jinja2Templates(directory=f'{BASE_PATH}/templates')

@app.get('/', response_class=HTMLResponse)
async def homepage(request: Request):
    return templates.TemplateResponse('index.html', {'request': request})
```

**Comment by ricardo-reis-1970** [comment]
# Problem: intercept the template generation process
I believe have the exact same problem as @siamhassan66. I too come from Flask. Here's what I think he meant: he wants to render the template into HTML *in the backend*, and get a result that is a string *with the HTML inside*. The documentation is strongly opinionated in this regard. It _assumes_ that all one could ever wish is to:
1. cook some values;
2. pass these to the `templates.TemplateResponse` call;
3. return immediately the outcome of `templates.TemplateResponse`.

Allow me to highlight a different scenario: I do not want to return that generated HTML, and for a good reason! My frontend is a Vue app. I am not serving pages there. All I do is communicate with the frontend via API calls and data.

What I really want here is to render a few templates, concatenate the results and conver them to PDF. Then — and only then! — do I want to return. Literally, you click on a button on the browser and seconds later here comes a PDF with a medical report.

# Flask and `render_template`
@siamhassan66 is absolutely right: in Flask, you do:
```python
from flask import render_template

html_fragment = render_template(
    'template.jinja2',
    param=value,
    another_parameter=another_value,
    i_think_you_are=starting_to_get_the_gist,
)
```
and that variable `html_fragment` is now a string with HTML inside. The way  have it, I collect a _variable_ number of HTML fragments and then I pass them to a final overall template.

In no step do In wish to return this. I pass the final full HTML document to a function based on `pdfkit`/`wkhtmltopdf`.

# Faslapi and `templates.TemplateResponse`

The only pattern present here, in all _fora_ and in the documentation is this:
```python
@app.get("/items/{id}", response_class=HTMLResponse)
async def read_item(request: Request, id: str):
    return templates.TemplateResponse("item.html", {"request": request, "id": id})
```

Everything is monolythically hardwired to the one-track process of returning the HTML.

Is it even possible to intercept this, as opposed to returning the HTML instantly? Is there _any_ method / property in `templates.TemplateResponse` that yields / holds the HTML?

If I find a solution, I'll post it here.

**Comment by siamhassan66** [comment]
it's been solved

**Comment by ricardo-reis-1970** [comment]
Dear @siamhassan66,

I realise you got somewhere, but I can't find _your_ solution here. I came here to find help and I'm sure more people do the same. So, as promised, here is the solution I came up with, hoping the next person will find it useful.

My approach is to break away from `templates.TemplateResponse`, because I came to realise that this is designed to do just that: pack and send back a response. Instead, I use a two-step approach where I acquire a template — via `templates.get_template` — and then I render it with some values — via `template.render`.

This approach allows me to get the generated HTML, while doing away with the need to provide a `request`. This way, the whole processing can be done at middleware level.

Here's the gist:
```python
from fastapi.templating import Jinja2Templates

templates = Jinja2Templates(directory=r"manager_backend\templates")
template = templates.get_template("item.jinja2")

@app.get("/items/{id}")
async def read_item(id: str):
    text = template.render(id=id)
    response = whatever(text) # do some processing with the text
    return response
```

Please notice that, compared with the previous snippet:
1. I changed the template from `item.html` to `item.jinja2`;
2. the arguments to `template.render` don't include the `request`, so I removed it from the path operation signature;
3. the arguments to `template.render` are no longer a `dict`, but rather `kwargs`;
4. I removed the `response_class=HTMLResponse` from the signature;
5. in my particular case, the function is returning a pdf file, but this is not a saved file, so I could not use `FileResponse`; instead, I did something like this:
```python
@router.get("/{id}")
def generate_report_id(id):
    pdf, file_name = generate_report(id)
    return Response(content = pdf, headers = {
        'Content-Type': 'application/pdf',
        'Content-Disposition': f'attachment; filename={file_name}',
        'x-suggested-filename': file_name,
    })
```
This allows me to click a button on the frontend and get a pdf downloaded.

I hope this helps. I wish I had something like this in the examples.
